migrating feature field  to array field

praffq opened the pull request:
## Proposed Changes

Migrations sequence -
- renamed feature to old_feature
- added a feature field with array field
- copied data from old_feature field to feature field
- deleted old_feature field

### Associated Issue

- fixes #2215 


## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

praffq committed:
migrating feature to array field

praffq committed:
updated serializer for feature field

praffq committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model

sainak commented on care/facility/api/serializers/facility.py:
```diff
@@ -95,7 +98,10 @@ class FacilitySerializer(FacilityBasicInfoSerializer):
     read_cover_image_url = serializers.URLField(read_only=True)
     # location = PointField(required=False)
-    features = serializers.MultipleChoiceField(choices=FEATURE_CHOICES)
+    features = serializers.ListField(
+        child=serializers.ChoiceField(choices=FEATURE_CHOICES),
+        required=False,
+    )
```
can you check if the validations are working 
- should not allow values not present in conices
- should not allow duplicates

sainak commented on care/facility/migrations/0438_rename_features_facility_old_features.py:
```diff
@@ -0,0 +1,51 @@
+def convert_features_to_array(apps, schema_editor):
+    Facility = apps.get_model("facility", "Facility")
+
+    for facility in Facility.objects.all():
+        if facility.old_features:
+            facility.features = list(map(int, facility.old_features))
+            facility.save()
```
use bulk update

sainak commented on care/facility/migrations/0438_rename_features_facility_old_features.py:
```diff
@@ -0,0 +1,51 @@
+            ),
+        ),
+        migrations.RunPython(convert_features_to_array),
+        migrations.RemoveField(
+            model_name="facility",
+            name="old_features",
+        ),
```
we can remove the field in a later pr

sainak commented on care/facility/models/facility.py:
```diff
@@ -160,11 +159,10 @@ class Facility(FacilityBaseModel, FacilityPermissionMixin):
-        choices=FEATURE_CHOICES,
-        null=True,
+    features = ArrayField(
+        models.SmallIntegerField(choices=FEATURE_CHOICES),
         blank=True,
-        max_length=get_max_length(FEATURE_CHOICES, None),
+        null=True,
     )
```
add the get_features_display method as the array field doesn't register this getter

praffq replied:
```
    @property
    def get_features_display(self):
        if not self.features:
            return []
        return [REVERSE_FEATURE_CHOICES[f] for f in self.features]
```

something like this ?

sainak replied:
yes

sainak reviewed (changes requested):
we can remove old_features in a later pr

praffq committed:
Added validations and tests

praffq committed:
Added tests

praffq committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model

praffq committed:
Merge remote-tracking branch 'origin/updating-feature-to-arrayfield-in-facility-model' into updating-feature-to-arrayfield-in-facility-model

praffq commented:
@sainak  made all the required changes, please take a look

sainak commented on care/facility/models/facility.py:
```diff
@@ -50,6 +51,7 @@
 ROOM_TYPES.extend(BASE_ROOM_TYPES)
 
 REVERSE_ROOM_TYPES = reverse_choices(ROOM_TYPES)
+REVERSE_FEATURE_CHOICES = reverse_choices(FEATURE_CHOICES)
```
we can use a new integer choice enum `FacilityFeature`
keep the old one in place, just replace it for the new field

praffq committed:
added integer choice enum for features

praffq committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model

praffq committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model

praffq committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model

sainak commented on care/facility/migrations/0438_rename_features_facility_old_features.py:
```diff
@@ -0,0 +1,54 @@
+
+    for facility in facilities_to_update:
+        facility.features = list(map(int, facility.old_features.split(",")))
+        updated_facilities.append(facility)
```
```suggestion
        try:
            facility.features = list(map(int, facility.old_features.split(",")))
            updated_facilities.append(facility)
        except ValueError:
            print(f"facility '{facility.name)' has invalid facility features")
```

sainak commented on care/facility/migrations/0438_rename_features_facility_old_features.py:
update the migration number and remove the merge migrations

praffq committed:
fixed migrations

praffq commented:
@sainak  @rithviknishad  made the changes , can you please run the workflows to check if everything's alright?

rithviknishad commented on care/facility/migrations/0445_rename_features_facility_old_features.py:
```diff
@@ -0,0 +1,56 @@
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("facility", "0437_alter_dailyround_rounds_type"),
```
this needs to be changed to the previous migration before it


```suggestion
        ("facility", "0444....."),
```

rithviknishad commented on care/facility/migrations/0446_merge_20240712_1357.py:
```diff
@@ -0,0 +1,13 @@
+# Generated by Django 4.2.10 on 2024-07-12 08:27
+
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("facility", "0444_alter_medicineadministration_dosage_and_more"),
+        ("facility", "0445_rename_features_facility_old_features"),
+    ]
+
+    operations = []
```
That way this merge migration can be deleted.

praffq committed:
deleted merge migration and updated the numbers

praffq committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model

praffq commented:
updated. Should be fine now.

praffq committed:
Merge branch 'coronasafe:develop' into updating-feature-to-arrayfield-in-facility-model

praffq committed:
fixed dummy data in facility.json

praffq committed:
fixed dummy data in facility.json

praffq commented:
@rithviknishad @sainak  I've fixed the errors , updated the migration file name and dependencies , also fixed the static dummy data from facility.json

praffq committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model

praffq committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model

praffq committed:
updated migration number

vigneshhari committed:
Merge branch 'develop' into updating-feature-to-arrayfield-in-facility-model
